modules/configuration/src/main/java/org/apache/ignite/configuration/ConfigurationChanger.java

```diff
@@ -115,50 +208,78 @@ private void change0(Map<RootKey<?>, TraversableTreeNode> changes, CompletableFu
             return;
         }
 
-        final int version = validationResult.version();
-
-        CompletableFuture<Boolean> writeFut = configurationStorage.write(allChanges, version);
+        CompletableFuture<Boolean> writeFut = storage.write(allChanges, roots.version);
 
         writeFut.whenCompleteAsync((casResult, throwable) -> {
             if (throwable != null)
                 fut.completeExceptionally(new ConfigurationChangeException("Failed to change configuration", throwable));
             else if (casResult)
                 fut.complete(null);
             else
-                change0(changes, fut);
+                change0(changes, storage, fut);
         }, pool);
     }
 
     /**
      * Update configuration from storage listener.
+     * @param storageType Type of the storage that propagated these changes.
      * @param changedEntries Changed data.
      */
-    private synchronized void updateFromListener(Data changedEntries) {
-        // TODO: IGNITE-14118 add tree update
-        version.set(changedEntries.version());
+    private void updateFromListener(
+        Class<? extends ConfigurationStorage> storageType,
+        Data changedEntries
+    ) {
+        StorageRoots oldStorageRoots = this.storagesRootsMap.get(storageType);
+
+        Map<RootKey<?>, InnerNode> storageRootsMap = new HashMap<>(oldStorageRoots.roots);
+
+        Map<String, ?> dataValuesPrefixMap = ConfigurationUtil.toPrefixMap(changedEntries.values());
+
+        for (RootKey<?> rootKey : oldStorageRoots.roots.keySet()) {
+            //TODO IGNITE-14182 Remove is not yet supported here.
+            Map<String, ?> rootPrefixMap = (Map<String, ?>)dataValuesPrefixMap.get(rootKey.key());
+
+            if (rootPrefixMap != null) {
+                InnerNode rootNode = oldStorageRoots.roots.get(rootKey).copy();
+
+                ConfigurationUtil.fillFromPrefixMap(rootNode, rootPrefixMap);
+
+                storageRootsMap.put(rootKey, rootNode);
+            }
+        }
+
+        StorageRoots storageRoots = new StorageRoots(storageRootsMap, changedEntries.version());
+
+        storagesRootsMap.put(storageType, storageRoots);
+
+        // Notify listeners.
```

Comment:
Should this have it's own todo maybe?

Reply:
Probably, I think I need to create an issue for that